File: src/main/java/org/taktik/icure/dao/impl/TimeTableDAOImpl.java
/*
 * Copyright (C) 2018 Taktik SA
 *
 * This file is part of iCureBackend.
 *
 * iCureBackend is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2 as published by
 * the Free Software Foundation.
 *
 * iCureBackend is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with iCureBackend.  If not, see <http://www.gnu.org/licenses/>.
 */

package org.taktik.icure.dao.impl;

import org.ektorp.ComplexKey;
import org.ektorp.ViewQuery;
import org.ektorp.support.View;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.beans.factory.annotation.Qualifier;
import org.springframework.stereotype.Repository;
import org.taktik.icure.dao.TimeTableDAO;
import org.taktik.icure.dao.impl.ektorp.CouchDbICureConnector;
import org.taktik.icure.dao.impl.idgenerators.IDGenerator;
import org.taktik.icure.entities.TimeTable;

import java.util.ArrayList;
import java.util.List;

@Repository("timeTableDAO")
@View(name = "all", map = "function(doc) { if (doc.java_type == 'org.taktik.icure.entities.TimeTable' && !doc.deleted) emit( null, doc._id )}")
public class TimeTableDAOImpl extends GenericDAOImpl<TimeTable> implements TimeTableDAO {

    @Autowired
    public TimeTableDAOImpl(@SuppressWarnings("SpringJavaAutowiringInspection") @Qualifier("couchdbHealthdata") CouchDbICureConnector db, IDGenerator idGenerator) {
        super(TimeTable.class, db, idGenerator);
        initStandardDesignDocument();
    }

    @Override
    @View(name = "by_agenda", map = "classpath:js/timeTable/by_agenda.js")
    public List<TimeTable> listTimeTableByAgendaId(String agendaId) {
        ViewQuery viewQuery = createQuery("by_agenda")
                .startKey(agendaId)
                .endKey(agendaId)
                .includeDocs(false);

        List<TimeTable> timeTables = db.queryView(viewQuery, TimeTable.class);

        return timeTables;
    }

    @Override
    @View(name = "by_agenda_and_startdate", map = "classpath:js/timeTable/by_agenda_and_startdate.js")
    public List<TimeTable> listTimeTableByStartDateAndAgendaId(Long startDate, Long endDate, String agendaId) {
        ComplexKey from = ComplexKey.of(
                agendaId,
                startDate
        );
        ComplexKey to = ComplexKey.of(
                agendaId,
                endDate == null ? ComplexKey.emptyObject() : endDate
        );

        ViewQuery viewQuery = createQuery("by_agenda_and_startdate")
                .startKey(from)
                .endKey(to)
                .includeDocs(false);

        List<TimeTable> timeTables = db.queryView(viewQuery, TimeTable.class);

        return timeTables;
    }

    @Override
    @View(name = "by_agenda_and_enddate", map = "classpath:js/timeTable/by_agenda_and_enddate.js")
    public List<TimeTable> listTimeTableByEndDateAndAgendaId(Long startDate, Long endDate, String agendaId) {
        ComplexKey from = ComplexKey.of(
                agendaId,
                startDate
        );
        ComplexKey to = ComplexKey.of(
                agendaId,
                endDate == null ? ComplexKey.emptyObject() : endDate
        );

        ViewQuery viewQuery = createQuery("by_agenda_and_enddate")
                .startKey(from)
                .endKey(to)
                .includeDocs(false);

        List<TimeTable> timeTables = db.queryView(viewQuery, TimeTable.class);

        return timeTables;
    }

    @Override
    public List<TimeTable> listTimeTableByPeriodAndAgendaId(Long startDate, Long endDate, String agendaId) {
        List<TimeTable> timeTablesStart = this.listTimeTableByStartDateAndAgendaId(startDate, endDate, agendaId);
        List<TimeTable> timeTablesEnd = this.listTimeTableByEndDateAndAgendaId(startDate, endDate, agendaId);
        /* Special case : timeTableStart < research.start < rechearch.end < timetableEnd*/
        List<TimeTable> timeTableStartBefore = this.listTimeTableByStartDateAndAgendaId(0l,startDate,agendaId);
        List<TimeTable> timeTableEndAfter = this.listTimeTableByEndDateAndAgendaId(endDate,999999999999999l,agendaId);
        List<TimeTable> timeTableMerged = new ArrayList<>();

        /* Add in merged TimeTable that are in both timeTableStartBefore AND timeTableEndAfter, avoiding duplicates*/
        for(TimeTable elem : timeTableStartBefore){
            if(listContains(timeTableEndAfter,elem) && !listContains(timeTableMerged,elem)){
                timeTableMerged.add(elem);
            }
        }
        /* Add in merged elem that are in timeTablesStart, avoiding duplicate */
        for(TimeTable elem : timeTablesStart){
            if(!listContains(timeTableMerged,elem)){
                timeTableMerged.add(elem);
            }
        }

        /* Add in merged elem that are in timeTablesEnd, avoiding duplicate */
        for(TimeTable elem : timeTablesEnd){
            if(!listContains(timeTableMerged,elem)){
                timeTableMerged.add(elem);
            }
        }
        return timeTableMerged;
    }

    private boolean listContains(List<TimeTable> list, TimeTable obj){
        for(TimeTable elem : list){
            if(elem.getId().equals(obj.getId())){
                return true;
            }
        }
        return false;
    }
}

File: src/main/kotlin/org/taktik/icure/asyncdao/impl/TimeTableDAOImpl.kt
/*
 *  iCure Data Stack. Copyright (c) 2020 Taktik SA
 *
 *     This program is free software: you can redistribute it and/or modify
 *     it under the terms of the GNU Affero General Public License as
 *     published by the Free Software Foundation, either version 3 of the
 *     License, or (at your option) any later version.
 *
 *     This program is distributed in the hope that it will be useful, but
 *     WITHOUT ANY WARRANTY; without even the implied warranty of
 *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 *     Affero General Public License for more details.
 *
 *     You should have received a copy of the GNU Affero General Public
 *     License along with this program.  If not, see
 *     <https://www.gnu.org/licenses/>.
 */

package org.taktik.icure.asyncdao.impl

import kotlinx.coroutines.flow.*
import org.taktik.couchdb.annotation.View
import org.springframework.beans.factory.annotation.Qualifier
import org.springframework.stereotype.Repository
import org.taktik.couchdb.entity.ComplexKey
import org.taktik.couchdb.queryViewIncludeDocs
import org.taktik.icure.asyncdao.TimeTableDAO
import org.taktik.couchdb.id.IDGenerator
import org.taktik.couchdb.queryViewIncludeDocsNoValue
import org.taktik.icure.entities.TimeTable
import org.taktik.icure.properties.CouchDbProperties
import org.taktik.icure.utils.FuzzyValues

import org.taktik.icure.utils.distinctById
import java.time.temporal.ChronoUnit

@Repository("timeTableDAO")
@View(name = "all", map = "function(doc) { if (doc.java_type == 'org.taktik.icure.entities.TimeTable' && !doc.deleted) emit( null, doc._id )}")
class TimeTableDAOImpl (couchDbProperties: CouchDbProperties,
                        @Qualifier("healthdataCouchDbDispatcher") couchDbDispatcher: CouchDbDispatcher, idGenerator: IDGenerator) : GenericDAOImpl<TimeTable>(couchDbProperties, TimeTable::class.java, couchDbDispatcher, idGenerator), TimeTableDAO {

	@View(name = "by_agenda", map = "classpath:js/timeTable/by_agenda.js")
	override fun listTimeTableByAgendaId(agendaId: String): Flow<TimeTable> = flow {
        val client = couchDbDispatcher.getClient(dbInstanceUrl)
        val viewQuery = createQuery(client, "by_agenda")
				.startKey(agendaId)
				.endKey(agendaId)
				.includeDocs(true)
        emitAll(client.queryViewIncludeDocsNoValue<String, TimeTable>(viewQuery).map{it.doc})
	}

	@View(name = "by_agenda_and_startdate", map = "classpath:js/timeTable/by_agenda_and_startdate.js")
	override fun listTimeTableByStartDateAndAgendaId(startDate: Long?, endDate: Long?, agendaId: String): Flow<TimeTable> = flow {
        val client = couchDbDispatcher.getClient(dbInstanceUrl)
        val from = ComplexKey.of(
				agendaId,
				startDate
		)
		val to = ComplexKey.of(
				agendaId,
				endDate ?: ComplexKey.emptyObject()
		)
		val viewQuery = createQuery(client, "by_agenda_and_startdate")
				.startKey(from)
				.endKey(to)
				.includeDocs(true)
        emitAll(client.queryViewIncludeDocsNoValue<Array<String>, TimeTable>(viewQuery).map{it.doc})
	}

	@View(name = "by_agenda_and_enddate", map = "classpath:js/timeTable/by_agenda_and_enddate.js")
	override fun listTimeTableByEndDateAndAgendaId(startDate: Long?, endDate: Long?, agendaId: String): Flow<TimeTable> = flow {
        val client = couchDbDispatcher.getClient(dbInstanceUrl)
        val from = ComplexKey.of(
				agendaId,
				startDate
		)
		val to = ComplexKey.of(
				agendaId,
				endDate ?: ComplexKey.emptyObject()
		)
		val viewQuery = createQuery(client, "by_agenda_and_enddate")
				.startKey(from)
				.endKey(to)
				.includeDocs(true)
        emitAll(client.queryViewIncludeDocsNoValue<Array<String>, TimeTable>(viewQuery).map{it.doc})
	}

	override fun listTimeTableByPeriodAndAgendaId(startDate: Long?, endDate: Long?, agendaId: String): Flow<TimeTable> =
	        listTimeTableByStartDateAndAgendaId(
                    startDate?.let {
                        /* 1 day in the past to catch long lasting events that could bracket the search period */
                        FuzzyValues.getFuzzyDateTime(FuzzyValues.getDateTime(it).minusDays(1), ChronoUnit.SECONDS)
                    },
                    endDate,
                    agendaId
            ).filter {
                it.endTime?.let { et -> et > (startDate ?: 0) } ?: true
            }
}
